# ML Project : Predicting Movie Revenue
Made by:

Corentin Maillard 21306

Mourad Mettioui 195019

## A) Data Understanding and analysis

### 1) Load the Dataset and libraries

In [152]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

dataset_file_path = 'dataset.csv'
dataset_supplement_file_path = 'dataset_supplement.csv'
# transform the first collum into an index
dataset_data = pd.read_csv(dataset_file_path, index_col=0)
dataset_supplement_data = pd.read_csv(dataset_supplement_file_path, index_col=0)

dataset_supplement_data.drop(columns=['title'], inplace=True)
left = dataset_data.set_index(['id'])
right = dataset_supplement_data.set_index(['movie_id'])
data_combind = left.join(right)
# data_combind.set_index(['id'])
data_combind = data_combind.sort_values(by = 'id')


### 2) Undersand the structure of the dataset

In [155]:
# examine the contents of the resultant Dataset using the head() command, which grabs the first five rows:
data_combind

vote_count  vote_average                        title   
id                                                              
5              530           6.5                   Four Rooms  \
11            6624           NaN                    Star Wars   
12            6122           7.6                 Finding Nemo   
13            7927           8.2                 Forrest Gump   
14            3313           7.9              American Beauty   
...            ...           ...                          ...   
426067           0           0.0             Midnight Cabaret   
426469           7           7.4             Growing Up Smith   
433715           0           NaN                       8 Days   
447027           0           NaN              Running Forever   
459488           0           0.0  To Be Frank, Sinatra at 100   

                                                  tagline    status   
id                                                                    
5       Twelve outrageous guests. Four scandalous requ...  Released  \
11           A long time ago in a galaxy far, far away...  Released   
12      There are 3.7 trillion fish in the ocean, they...  Released   
13      The world will never be the same, once you've ...  Released   
14                                           Look closer.  Released   
...                                                   ...       ...   
426067                The hot spot where Satan's waitin'.  Released   
426469           It’s better to stand out than to fit in.  Released   
433715           She never knew it could happen to her...  Released   
447027                                                NaN  Released   
459488                                                NaN  Released   

                                spoken_languages  duration      revenue   
id                                                                        
5       [{"iso_639_1": "en", "name": "English"}]      98.0    4300000.0  \
11      [{"iso_639_1": "en", "name": "English"}]     121.0  775398007.0   
12      [{"iso_639_1": "en", "name": "English"}]     100.0  940335536.0   
13      [{"iso_639_1": "en", "name": "English"}]     142.0  677945399.0   
14      [{"iso_639_1": "en", "name": "English"}]     122.0  356296601.0   
...                                          ...       ...          ...   
426067  [{"iso_639_1": "en", "name": "English"}]      94.0          0.0   
426469  [{"iso_639_1": "en", "name": "English"}]     102.0          0.0   
433715  [{"iso_639_1": "en", "name": "English"}]      90.0          0.0   
447027                                        []      88.0          0.0   
459488                                        []       NaN          0.0   

           release countries_of_production  ...  popularity   
id                                          ...               
5       1995-12-09                      US  ...   22.876230  \
11      1977-05-25                      US  ...  126.393695   
12      2003-05-30                      US  ...   85.688789   
13      1994-07-06                      US  ...  138.133331   
14      1999-09-15                      US  ...   80.878605   
...            ...                     ...  ...         ...   
426067  1990-01-01                          ...    0.001389   
426469  2017-02-03                          ...    0.710870   
433715  2014-06-15                      US  ...    0.015295   
447027  2015-10-27                      US  ...    0.028756   
459488  2015-12-12                      GB  ...    0.050625   

                                                 overview   
id                                                          
5       It's Ted the Bellhop's first night on the job....  \
11      Princess Leia is captured and held hostage by ...   
12      Nemo, an adventurous young clownfish, is unexp...   
13      A man with a low IQ has accomplished great thi...   
14      Lester Burnham, a depressed suburban father in...   
...            

We can see the different features which are:
- vote count
- vote average
- title
- tagline
- status
- spoken languages
- duration
- release
- countries_of_production
- production
- popularity
- overview
- original_title
- original_language
- keywords
- id
- homepage
- genres
- financial_investment
- actors
- production_crew

And we have the target : 
- revenue

In [123]:
# we are going to see if all the data that we have are complete
data_combind.info()
# we can also use the isnull().sum() to see directly how many data are missing
data_combind.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 4803 entries, 5 to 459488
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   vote_count               4803 non-null   int64  
 1   vote_average             4323 non-null   float64
 2   title                    4803 non-null   object 
 3   tagline                  3959 non-null   object 
 4   status                   4803 non-null   object 
 5   spoken_languages         4803 non-null   object 
 6   duration                 4801 non-null   float64
 7   revenue                  4755 non-null   float64
 8   release                  4802 non-null   object 
 9   countries_of_production  4803 non-null   object 
 10  production               4803 non-null   object 
 11  popularity               4803 non-null   float64
 12  overview                 4800 non-null   object 
 13  original_title           4803 non-null   object 
 14  original_language        45

vote_count                    0
vote_average                480
title                         0
tagline                     844
status                        0
spoken_languages              0
duration                      2
revenue                      48
release                       1
countries_of_production       0
production                    0
popularity                    0
overview                      3
original_title                0
original_language           240
keywords                      0
homepage                   3091
genres                        0
financial_investment        384
actors                        0
production_crew               0
dtype: int64

We can see that they are only 1712 non-null homepage on 4803 entires, so we will discard this feature because they are too many missing datas.

For the other features, we can see if they are complete or not

In [139]:
# we are going to see if all the data are different or not
data_combind.nunique()

vote_count                 1609
vote_average                 70
title                      4800
tagline                    3944
status                        3
spoken_languages            544
duration                    156
revenue                    3268
release                    3280
countries_of_production     469
production                 3697
popularity                 4802
overview                   4800
original_title             4801
original_language            37
keywords                   4222
homepage                   1691
genres                     1175
financial_investment        412
actors                     4761
production_crew            4776
dtype: int64

In [135]:
# we will drop the id because it is not an information that will ever be seen by the custommer.
# So we can assume that it is not an important feature for our model.
# data_combind.drop(columns=['id'], inplace=True)
data_combind.drop(columns=['homepage'], inplace=True)
data_combind.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4803 entries, 5 to 459488
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   vote_count               4803 non-null   int64  
 1   vote_average             4323 non-null   float64
 2   title                    4803 non-null   object 
 3   tagline                  3959 non-null   object 
 4   status                   4803 non-null   object 
 5   spoken_languages         4803 non-null   object 
 6   duration                 4801 non-null   float64
 7   revenue                  4755 non-null   float64
 8   release                  4802 non-null   object 
 9   countries_of_production  4803 non-null   object 
 10  production               4803 non-null   object 
 11  popularity               4803 non-null   float64
 12  overview                 4800 non-null   object 
 13  original_title           4803 non-null   object 
 14  original_language        45

### 3) Perform Exploratory Data Analysis (EDA)

In [147]:
data_combind.describe()

vote_count  vote_average     duration       revenue   popularity   
count   4803.000000   4323.000000  4801.000000  4.755000e+03  4803.000000  \
mean     690.217989      6.090354   106.875859  8.231486e+07    21.492301   
std     1234.585891      1.193315    22.611935  1.630872e+08    31.816650   
min        0.000000      0.000000     0.000000  0.000000e+00     0.000000   
25%       54.000000      5.600000    94.000000  0.000000e+00     4.668070   
50%      235.000000      6.200000   103.000000  1.918402e+07    12.921594   
75%      737.000000      6.800000   118.000000  9.311911e+07    28.313505   
max    13752.000000     10.000000   338.000000  2.787965e+09   875.581305   

       financial_investment  
count          4.419000e+03  
mean           2.898466e+07  
std            4.065526e+07  
min            0.000000e+00  
25%            7.500000e+05  
50%            1.480000e+07  
75%            4.000000e+07  
max            3.800000e+08

In [161]:
# sns.pairplot(dataset_data)  # Visualize pairwise relationships using a pairplot
# plt.show()

We can see that that the revenue will be affected by:
- the financial investment
- the vote count

For the other features (vote average, duration, popularity), it is not really visible the impact they have an the revenue

We can see that the vote average is affected by:
- the vote count
- the financial investment

We can see a big problem, that is the feature that are not number cannot be used to make an EDA with the way I make it.

We will use the one hot encoding approach that generally perform best on:
- the country_of_production
- the spoken_languages
- genres
- status
- original_language

For the feature like the title, tagline, overview, original_title we know that they can have an influence on the renenue but they are to complexe to exploit so we will discard them

In [163]:

import json

def extract_iso(df, col_name, iso_num):
    collone_brut = df[col_name]
    
    for i in df.index:
        langperligne = []
        lst = json.loads(collone_brut[i])
        for item in lst:
            langperligne.append(item[iso_num])
        collone_brut[i] = ','.join(langperligne)
    
    return collone_brut

# Utilisez la fonction avec le nom de la colonne et le numéro d'ISO

# extract_iso(data_combind, 'countries_of_production', 'iso_3166_1')
# extract_iso(data_combind, 'spoken_languages', 'iso_639_1')
data_combind

# print(data_combind.nunique())

CA,ES


vote_count  vote_average                        title   
id                                                              
5              530           6.5                   Four Rooms  \
11            6624           NaN                    Star Wars   
12            6122           7.6                 Finding Nemo   
13            7927           8.2                 Forrest Gump   
14            3313           7.9              American Beauty   
...            ...           ...                          ...   
426067           0           0.0             Midnight Cabaret   
426469           7           7.4             Growing Up Smith   
433715           0           NaN                       8 Days   
447027           0           NaN              Running Forever   
459488           0           0.0  To Be Frank, Sinatra at 100   

                                                  tagline    status   
id                                                                    
5       Twelve outrageous guests. Four scandalous requ...  Released  \
11           A long time ago in a galaxy far, far away...  Released   
12      There are 3.7 trillion fish in the ocean, they...  Released   
13      The world will never be the same, once you've ...  Released   
14                                           Look closer.  Released   
...                                                   ...       ...   
426067                The hot spot where Satan's waitin'.  Released   
426469           It’s better to stand out than to fit in.  Released   
433715           She never knew it could happen to her...  Released   
447027                                                NaN  Released   
459488                                                NaN  Released   

       spoken_languages  duration      revenue     release   
id                                                           
5                    en      98.0    4300000.0  1995-12-09  \
11                   en     121.0  775398007.0  1977-05-25   
12                   en     100.0  940335536.0  2003-05-30   
13                   en     142.0  677945399.0  1994-07-06   
14                   en     122.0  356296601.0  1999-09-15   
...                 ...       ...          ...         ...   
426067               en      94.0          0.0  1990-01-01   
426469               en     102.0          0.0  2017-02-03   
433715               en      90.0          0.0  2014-06-15   
447027                       88.0          0.0  2015-10-27   
459488                        NaN          0.0  2015-12-12   

       countries_of_production  ...  popularity   
id                              ...               
5                           US  ...   22.876230  \
11                          US  ...  126.393695   
12                          US  ...   85.688789   
13                          US  ...  138.133331   
14                          US  ...   80.878605   
...                        ...  ...         ...   
426067                          ...    0.001389   
426469                          ...    0.710870   
433715                      US  ...    0.015295   
447027                      US  ...    0.028756   
459488                      GB  ...    0.050625   

                                                 overview   
id                                                          
5       It's Ted the Bellhop's first night on the job....  \
11      Princess Leia is captured and held hostage by ...   
12      Nemo, an adventurous young clownfish, is unexp...   
13      A man with a low IQ has accomplished great thi...   
14      Lester Burnham, a depressed suburban father in...   
...                                                   ...   
426067  A Broadway producer puts on a play with a Devi...   
426469  In 1979, an Indian family moves to America wit...   
433715  After sneaking to a party with her friends, 16...   
447027  After being estranged since her mother's death...   
459488                             